# Deep Learning: 

# Homework 3

Your task is to predict whether a patient has heart disease or not. You can
download and see information about data by following this link. You should do
the following steps in your homework:

    a) Keep 20 percent of the data for testing.  
    b) Do logistic regression and try to find the best hyperparameters (usingsklearn).
    c) Normalize your data using standard normalization and then repeat theprevious step.
    d) Try to find the best neural net to solve this problem which will have nomore than 2 hidden layers (use sklearn).
    e) Normalize your data using standard normalization and then repeat theprevious step.
    f*) Do the step d) using tf.keras.
    g) Compare accuracies of all obtained models.

### 1. Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score

# Normalization
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler

### 2. Data Exploration

In [2]:
heart = pd.read_csv('./data/heart.csv')
heart.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [3]:
heart.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
age         303 non-null int64
sex         303 non-null int64
cp          303 non-null int64
trestbps    303 non-null int64
chol        303 non-null int64
fbs         303 non-null int64
restecg     303 non-null int64
thalach     303 non-null int64
exang       303 non-null int64
oldpeak     303 non-null float64
slope       303 non-null int64
ca          303 non-null int64
thal        303 non-null int64
target      303 non-null int64
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [4]:
heart.shape

(303, 14)

In [5]:
heart.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

### 3. Model implementation

#### 3.1 Logistic Regression

(3.1.1 without normalization)

In [58]:
# Initialize model
logreg = LogisticRegression()

In [59]:
# Train-test split

X_train, X_test, y_train, y_test = train_test_split(heart.drop(['target'], axis = 1), heart['target'], test_size=0.20, 
                                                    random_state=42)

In [60]:
logreg.fit(X_train, y_train)
y_predict = logreg.predict(X_test)

accuracy_score(y_test, y_predict)

C:\Users\Lilit\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.8852459016393442

(3.1.2 with normalization)

In [61]:
norm_X_train = (X_train-np.min(X_train))/(np.max(X_train)-np.min(X_train)).values

norm_X_test = (X_test-np.min(X_test))/(np.max(X_test)-np.min(X_test)).values

In [62]:
logreg.fit(norm_X_train, y_train)
Y_predict_trial = logreg.predict(norm_X_test)
accuracy_score(y_test, Y_predict_trial)

0.8524590163934426

In [63]:
# NOTE: Here I used normalization techniques due to which I didn't get a convergence warning but the accuracy score
# decreases, which is strange, since I was expecting to see the inverse picture. I tried to change the max_iter default
# number, but this removes the warning message only, and the main "question" stays.

In [11]:
# norm_X_train = preprocessing.normalize(X_train)
# norm_X_train = min_max_scaler.fit_transform(X_train)
# norm_X_train = preprocessing.scale(X_train)
# norm_X_train = preprocessing.StandardScaler().fit(X_train)
# 
# norm_X_test = preprocessing.normalize(X_test)
# norm_X_test = min_max_scaler.fit_transform(X_test)
# norm_X_test = preprocessing.scale(X_test)
# norm_X_test = preprocessing.StandardScaler().fit(X_test)

#### 3.2 Neural Network

(3.2.1 without normalization)

In [30]:
model = MLPClassifier(hidden_layer_sizes=(2,2), max_iter=150000)
model.fit(X_train,y_train)

ypred = model.predict(X_test)

In [31]:
acc = accuracy_score(y_test,ypred)
print(acc*100)

85.24590163934425


(3.2.2 with normalization)

In [32]:
model.fit(norm_X_train, y_train)
Ypred = model.predict(norm_X_test)

acc = accuracy_score(y_test,Ypred)
print(acc*100)

86.88524590163934


For the Logistic Regression "without normalization" case I got 88.5% accuracy, but meanwhile I got a convergence warning, after normalization that warning disappeared but the accuracy decreased to 85.24%. For the Neural Network 2 hidden layer and "without normalization" case I got 85.24% accuracy, after normalizing that score increased to 86.89%